decentraland/unity-explorer
# fix: map pins systems refactor for proper cleanup

## What does this PR change?

The issue was twofold, in one hand, we were not calling the Finalize World Systems when a PX was killed and in the other hand we were not properly cleaning map pins when we removed their scene counterpart.
The fix for the first was adding calls to Finalize World Systems when we are killing the PXs and adding that logic also to the MapPins system so its called when we do the finalize worlds.

The other change was a partial refactor of how we setup and update map pins, instead of using a cloned entity in the global world and in the scene world, I opted for using a bus (as we have used for other UI facing systems). This way the UI can react to the ECS changes instantly and without having to rely on another system running in parallel to update the UI.
This both simplifies creating and deletion of map pins and also makes it easier to track.
...

## How to test the changes?

This fixes this issue: https://github.com/decentraland/unity-explorer/issues/3023
Also fixes the issue that would happen if we killed the global portable experience while we still had missions available. We would still see the map pins on the map even if the px was not running.

To test this there are 2 ways:
* Enter the game with one user, wait until the missions are loaded and see that you have the map pins on the map. Log out and log back in with another user that has different missions. You should only see the map pins of the new user and not the ones of the previous one. Ç
If both users have the same missions, an easy way to see if this is working is to complete one mission with one of the accounts, so you only have 2 map pins on the map. When you log in with the other account you should see all the map pins instead of only 2.

* The second test is to run the command /killpx globalpx, this should kill the portable experience, making the rocket icon on the top disappear and also the map pins on the map. If you then run /loadpx globalpx, the map pins should appear again in the map as before.
